// geojson-path-finder/dist/cjs/topology.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultKey = void 0;
const helpers_1 = require("@turf/helpers");
const explode_1 = __importDefault(require("@turf/explode"));
const round_coord_1 = __importDefault(require("./round-coord"));
function createTopology(network, options = {}) {
    const { key = defaultKey } = options;
    const { tolerance = 1e-5 } = options;
    const lineStrings = (0, helpers_1.featureCollection)(network.features.filter((f) => f.geometry.type === "LineString"));
    const points = (0, explode_1.default)(lineStrings);
    const vertices = points.features.reduce(function buildTopologyVertices(coordinates, feature, index, features) {
        var rc = (0, round_coord_1.default)(feature.geometry.coordinates, tolerance);
        coordinates[key(rc)] = feature.geometry.coordinates;
        if (index % 1000 === 0 && options.progress) {
            options.progress("topo:vertices", index, features.length);
        }
        return coordinates;
    }, {});
    const edges = geoJsonReduce(lineStrings, buildTopologyEdges, []);
    return {
        vertices: vertices,
        edges: edges,
    };
    function buildTopologyEdges(edges, f) {
        f.geometry.coordinates.forEach(function buildLineStringEdges(c, i, cs) {
            if (i > 0) {
                var k1 = key((0, round_coord_1.default)(cs[i - 1], tolerance)), k2 = key((0, round_coord_1.default)(c, tolerance));
                edges.push([k1, k2, f.properties]);
            }
        });
        return edges;
    }
}
exports.default = createTopology;
function geoJsonReduce(geojson, fn, seed) {
    if (geojson.type === "FeatureCollection") {
        return geojson.features.reduce(function reduceFeatures(a, f) {
            return geoJsonReduce(f, fn, a);
        }, seed);
    }
    else {
        return fn(seed, geojson);
    }
}
function defaultKey(c) {
    return c.join(",");
}
exports.defaultKey = defaultKey;
